pilot/adapter/config/file/monitor.go

```diff
@@ -0,0 +1,390 @@
+	kinds         map[string]bool
+	checkDuration time.Duration
+	resources     resources
+}
+
+func (m *Monitor) checkAndUpdate() {
+	// Read all of the files under the directory and sort them based on their kind, name, and namespace.
+	newResources := m.readFiles()
+
+	// Compare the new list to the previous one and detect changes.
+	oldLen := len(m.resources)
+	newLen := len(newResources)
+	oldIndex, newIndex := 0, 0
+	for oldIndex < oldLen && newIndex < newLen {
+		oldResource := m.resources[oldIndex]
+		newResource := newResources[newIndex]
+		if v := compareResourceIds(oldResource, newResource); v < 0 {
+			m.deleteConfig(oldResource)
+			oldIndex++
+		} else if v > 0 {
+			m.createConfig(newResource)
+			newIndex++
+		} else {
+			if oldResource.sha != newResource.sha {
+				m.updateConfig(newResource)
+			}
+			oldIndex++
+			newIndex++
+		}
+	}
+
+	// Detect remaining deletions
+	for ; oldIndex < oldLen; oldIndex++ {
+		m.deleteConfig(m.resources[oldIndex])
+	}
+
+	// Detect remaining additions
+	for ; newIndex < newLen; newIndex++ {
+		m.createConfig(newResources[newIndex])
+	}
+
+	// Save the updated list.
+	m.resources = newResources
+}
+
+func (m *Monitor) createConfig(r *resource) {
+	cfg, err := r.toConfig()
+	if err != nil {
+		glog.Warningf("Failed to parse config (%m): %v", r.Metadata, err)
+		return
+	}
+
+	if _, err = m.store.Create(*cfg); err != nil {
+		glog.Warningf("Failed to create config (%m): %v ", r.Metadata, err)
+	}
+}
+
+func (m *Monitor) updateConfig(r *resource) {
+	cfg, err := r.toConfig()
+	if err != nil {
+		glog.Warningf("Failed to parse config (%m): %v", r.Metadata, err)
+		return
+	}
+
+	// Set the resource version based on the existing resource.
+	if prev, exists := m.store.Get(cfg.Type, cfg.Name, cfg.Namespace); exists {
+		cfg.ResourceVersion = prev.ResourceVersion
+	}
+
+	if _, err = m.store.Update(*cfg); err != nil {
+		glog.Warningf("Failed to update config (%m): %v ", r.Metadata, err)
+	}
+}
+
+func (m *Monitor) deleteConfig(r *resource) {
+	schema, ok := schemaForKind(r.Kind)
+	if !ok {
+		glog.Warningf("Missing schema for %q ", r.Kind)
+		return
+	}
+
+	if err := m.store.Delete(schema.Type, r.Metadata.Name, r.Metadata.Namespace); err != nil {
+		glog.Warningf("Failed to delete config (%m): %v ", r.Metadata, err)
+	}
+}
+
+func (m *Monitor) readFiles() resources {
+	var result resources
+
+	err := filepath.Walk(m.root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !supportedExtensions[filepath.Ext(path)] || (info.Mode()&os.ModeType) != 0 {
+			return nil
+		}
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			glog.Warningf("Failed to read %m: %v", path, err)
+			return err
+		}
+		for _, r := range parseYaml(data) {
+			if !m.kinds[r.Kind] {
+				// Ignore unsupported kinds
+				continue
+			}
+			result = append(result, r)
+		}
+		return nil
+	})
+	if err != nil {
+		glog.Errorf("failure during filepath.Walk: %v", err)
```

Comment:
Keep this a warning - Error indicates some irrecoverable thing happened on the server and typically implies that the server will stop serving afterwards.